uast/diff/diff.go

```diff
@@ -0,0 +1,287 @@
+package diff
+
+import (
+	"fmt"
+	"github.com/heetch/lapjv"
+	"gopkg.in/bblfsh/sdk.v2/uast/nodes"
+)
+
+type decisionType interface {
+	isDecisionType()
+	cost() int
+}
+
+type basicDecision struct {
+	privateCost int
+}
+
+func (basicDecision) isDecisionType() {}
+
+func (b basicDecision) cost() int { return b.privateCost }
+
+// match decision types together with their params
+type sameDecision struct{ basicDecision }
+type replaceDecision struct{ basicDecision }
+type matchDecision struct{ basicDecision }
+type permuteDecision struct {
+	basicDecision
+	permutation []int
+}
+
+//convinience method for choosing the cheapest decision
+func min(self, candidate decisionType) decisionType {
+	if self.cost() > candidate.cost() {
+		return candidate
+	} else {
+		return self
+	}
+}
+
+//type for cache
+type keyType struct{ k1, k2 ID }
+
+//cache for diff computation
+type cacheStorage struct {
+	decisions map[keyType]decisionType
+	sizes     map[ID]int
+	changes   Changelist
+}
+
+func emptyCacheStorage() cacheStorage {
+	return cacheStorage{
+		make(map[keyType]decisionType),
+		make(map[ID]int),
+		nil,
```

Comment:
Again, please don't initialize things with `nil`. It's not C - Go takes care of memory initialization with zero values.
Use named fields instead:
```
&cacheStorage{
  decisions: make(map[keyType]decisionType),
  sizes: make(map[ID]int),
}
```